azurerm/internal/services/datamigration/resource_arm_data_migration_service.go

```diff
@@ -0,0 +1,247 @@
+			}
+		}
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_data_migration_service", *existing.ID)
+		}
+	}
+
+	skuName := d.Get("sku_name").(string)
+	virtualSubnetID := d.Get("virtual_subnet_id").(string)
+	location := d.Get("location").(string)
+
+	parameters := datamigration.Service{
+		Location: utils.String(location),
+		ServiceProperties: &datamigration.ServiceProperties{
+			VirtualSubnetID: utils.String(virtualSubnetID),
+		},
+		Sku: &datamigration.ServiceSku{Name: utils.String(skuName)},
+	}
+	if kind, ok := d.GetOk("kind"); ok {
+		parameters.Kind = utils.String(kind.(string))
+	}
+	if t, ok := d.GetOk("tags"); ok {
+		parameters.Tags = tags.Expand(t.(map[string]interface{}))
+	}
+
+	future, err := client.CreateOrUpdate(ctx, parameters, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error creating Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for creation of Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Data Migration Service (Service Name %q / Group Name %q) ID", name, resourceGroup)
+	}
+	d.SetId(*resp.ID)
+
+	return resourceArmDataMigrationServiceRead(d, meta)
+}
+
+func resourceArmDataMigrationServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataMigration.ServicesClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["services"]
+
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[INFO] Data Migration Service %q does not exist - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Data Migration Service (Service Name %q / Group Name %q): %+v", name, resourceGroup, err)
+	}
+
+	d.Set("id", resp.ID)
+	if location := resp.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+	d.Set("resource_group_name", resourceGroup)
+	d.Set("name", name)
+	if serviceProperties := resp.ServiceProperties; serviceProperties != nil {
+		d.Set("provisioning_state", string(serviceProperties.ProvisioningState))
+		d.Set("virtual_subnet_id", serviceProperties.VirtualSubnetID)
+	}
+	if err := d.Set("sku_name", resp.Sku.Name); err != nil {
+		return fmt.Errorf("Error setting `sku_name`: %+v", err)
+	}
```

Comment:
we don't need to check for an error here, should case name to a string, and need to nil check SKU

Reply:
I will nil check resp.Sku. While for the casting, I think d.Set() already handle the case where you are passing in a pointer to a primitive type (resp.Sku.Name here is of type *string) and seems not necessary. Or no?